File: 백준/Gold/9251. LCS/LCS.java
import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

	public static void main(String[] args) throws IOException {
//		System.setIn(new FileInputStream("test.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		String s1 = br.readLine();
		String s2 = br.readLine();
		int l1 = s1.length();
		int l2 = s2.length();

		int dp[][] = new int[l2 + 1][l1 + 1];

		for (int j = 1; j <= l1; j++) {
			for (int i = 1; i <= l2; i++) {
				int temp = Math.max(dp[i-1][j], dp[i][j-1]);
				int temp2=0;
				if(s1.charAt(j-1) == s2.charAt(i-1)) {
					temp2 = dp[i-1][j-1]+1;
				}
				dp[i][j] = Math.max(temp, temp2);
			}
		}
		System.out.println(dp[l2][l1]);
	}
}

File: 백준/Gold/9251. LCS/LCS.cc
#include<iostream>
#include<string>

using namespace std;

int LCS[1001][1001];

int main() {
	string s1, s2;
	cin >> s1 >> s2;
	for (int i = 0;i <= s2.size();i++) {
		LCS[i][0] = 0;
	}
	for (int j = 0;j <= s1.size();j++) {
		LCS[0][j] = 0;
	}
	for (int i = 0;i < s2.size();i++) {
		for (int j = 0;j < s1.size();j++) {
			int temp = max(LCS[i+1][j], LCS[i][j + 1]);
			temp = max(LCS[i][j] + (s1[j] == s2[i]), temp);
			LCS[i + 1][j + 1] = temp ;
		}
	}
	cout << LCS[s2.size()][s1.size()];

} 